### PR TITLE
[ISSUE-364] Implement controller for drive and lvg and refactor discoverVolumeCRs 

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
### Issue #364 

Add proposal about refactoring logic with discovering if disk is clean or dirty and logic with AvailableCapacity manipulation in controller side

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [x] New unit tests added
- [x] Modified code has meaningful comments
- [x] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
_Link to custom CI build_

### CODE REVIEW

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+func (d *Controller) reconcileDrive(ctx context.Context, drive *drivecrd.Drive) (ctrl.Result, error) {
+	var (
+		health  = drive.Spec.GetHealth()
+		status  = drive.Spec.GetStatus()
+		isClean = drive.Spec.GetIsClean()
+	)
+	switch {
+	case isClean && (health == apiV1.HealthGood && status == apiV1.DriveStatusOnline):
+		return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
+	case health != apiV1.HealthGood || status != apiV1.DriveStatusOnline:
```
Comment (safronovD) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:130]: Is this not the same as
```
switch {
case health != apiV1.HealthGood || status != apiV1.DriveStatusOnline:
  return d.handleDriveIsNotGood(ctx, drive.Spec)
default:
  return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
```
Or we have some kind of special cases here?

File: pkg/node/volumemgr.go [hunk 2]
```diff
@@ -1036,16 +938,9 @@ func (m *VolumeManager) handleDriveStatusChange(ctx context.Context, drive *api.
...
-		if ac, err := m.cachedCrHelper.GetACByLocation(drive.UUID); err == nil {
-			ll.Infof("Removing AC %s based on unhealthy location %s", ac.Name, ac.Spec.Location)
-			if err := m.k8sClient.DeleteCR(ctx, ac); err != nil {
-				ll.Errorf("Failed to delete unhealthy available capacity CR: %v", err)
-			}
-		}
-	}
 	lvg, err := m.cachedCrHelper.GetLVGByDrive(ctx, drive.UUID)
 	if lvg != nil {
+		// TODO handle situation when LVG health is changing from Bad/Suspec to Good
```
Comment (safronovD) [inline pkg/node/volumemgr.go:943]: Issue?

**Review by safronovD — APPROVED** [review]

File: cmd/controller/main.go [hunk 3]
```diff
@@ -142,19 +143,14 @@ func main() {
...
+func createReservationManager(client *k8s.KubeClient, log *logrus.Logger, ch <-chan struct{}) (manager.Manager, error) {
 	// create scheme
-	scheme := runtime.NewScheme()
-	if err := clientgoscheme.AddToScheme(scheme); err != nil {
-		return nil, err
-	}
-
-	// register ACR CRD
-	if err = acrcrd.AddToSchemeACR(scheme); err != nil {
+	scheme, err := k8s.PrepareScheme()
```
Comment (mishoyama) [inline cmd/controller/main.go:148]: Why do we need to register for all CRDs?

File: cmd/controller/main.go [hunk 4]
```diff
@@ -142,19 +143,14 @@ func main() {
 	logger.Info("Got SIGTERM signal")
 }
 
-func createReservationManager(client *k8s.KubeClient, log *logrus.Logger) (mgr ctrl.Manager, err error) {
+func createReservationManager(client *k8s.KubeClient, log *logrus.Logger, ch <-chan struct{}) (manager.Manager, error) {
```
Comment (mishoyama) [inline cmd/controller/main.go:146]: Please rename this method

File: cmd/controller/main.go [hunk 5]
```diff
@@ -168,6 +164,17 @@ func createReservationManager(client *k8s.KubeClient, log *logrus.Logger) (mgr c
 	if err = reservationController.SetupWithManager(mgr); err != nil {
 		return nil, err
 	}
+	wrappedK8SClient := k8s.NewKubeClient(client, log, *namespace)
 
+	kubeCache, err := k8s.InitKubeCache(log, ch,
```
Comment (mishoyama) [inline cmd/controller/main.go:169]: Can we use cached client?

File: pkg/base/linuxutils/fs/fs_helper.go [hunk 6]
```diff
@@ -312,3 +315,24 @@ func (h *WrapFSImpl) Unmount(path string) error {
...
+func (h *WrapFSImpl) DriveHasData(device string) (bool, error) {
+	var (
+		cmd          = fmt.Sprintf(DetectFSCmdTmpl, device)
+		stdout       string
+		driveHasData bool
+		err          error
+	)
+	if stdout, _, err = h.e.RunCmd(cmd,
+		command.UseMetrics(true),
+		command.CmdName(strings.TrimSpace(fmt.Sprintf(DetectFSCmdTmpl, "")))); err != nil {
```
Comment (mishoyama) [inline pkg/base/linuxutils/fs/fs_helper.go:331]: You need to check not only FS, but:
1. Partition table
2. Partitions
3. FS on disk or FS signature
4. Disk is a PV for LVG

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+	ctx, cancelFn := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancelFn()
+
+	log := d.log.WithFields(logrus.Fields{"method": "Reconcile", "name": resourceName})
+
+	drive := &drivecrd.Drive{}
+	if err := d.client.ReadCR(ctx, resourceName, "", drive); err == nil {
+		return d.reconcileDrive(ctx, drive)
+	}
+	log.Warnf("Failed to read Drive %s CR, try to read LVG CR", resourceName)
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:84]: Please check for not found error and print warning when all resources not found.

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+		health = lvg.Spec.GetHealth()
+		name   = lvg.GetName()
+	)
+	// If LVG status is failed or Health is not good, try to reset its AC size to 0
+	if status == apiV1.Failed || health != apiV1.HealthGood {
+		return ctrl.Result{}, d.resetACSizeOfLVG(name)
+	}
+	// If LVG is already presented on a machine but doesn't have AC, try to create its AC using annotation with
+	// VG free space
+	size, err := getFreeSpaceFromLVGAnnotation(lvg.Annotations)
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:109]: Why do we need to use LVG annotations when original size is known and you can read related volumes info? Is that for SYSLVG?

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+		// If ac is exists, update its size to drive size
+		if ac.Spec.Size != size {
+			ac.Spec.Size = size
+			if err := d.client.Update(context.WithValue(ctx, base.RequestUUID, ac.Name), ac); err != nil {
+				log.Errorf("Error during update AvailableCapacity request to k8s: %v, error: %v", ac, err)
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{RequeueAfter: RequeueDriveTime}, nil
+	case err == errTypes.ErrorNotFound:
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:173]: Is that possible? Before we create LVG we modify AC type...

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+		}
+	default:
+		log.Infof("Failed to read AvailableCapacity for drive %s: %v", driveUUID, err)
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{RequeueAfter: RequeueDriveTime}, nil
+}
+
+// handleDriveIsNotGood deletes AC for bad Drive
+func (d *Controller) handleDriveIsNotGood(ctx context.Context, drive api.Drive) (ctrl.Result, error) {
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:198]: Please rename to `handleInaccessibleDrive`

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+	log := d.log.WithFields(logrus.Fields{
+		"method": "handleDriveIsNotGood",
+	})
+	ac, err := d.cachedCrHelper.GetACByLocation(drive.GetUUID())
+	switch {
+	case err == nil:
+		log.Infof("Removing AC %s based on unhealthy location %s", ac.Name, ac.Spec.Location)
+		if err := d.client.DeleteCR(ctx, ac); err != nil {
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:206]: You need to set size to 0 - not remove

**Review by mishoyama — CHANGES_REQUESTED** [review]

**Comment by mishoyama** [comment]
@AnastasiaShemyakinskaya  I added first portion of comments. Still looking at remaining changes...

File: pkg/controller/controller.go [hunk 12]
```diff
@@ -90,6 +90,7 @@ func NewControllerService(k8sClient *k8s.KubeClient, logger *logrus.Logger,
 // Probe is the implementation of CSI Spec Probe for IdentityServer.
 // This method checks if CSI driver is ready to serve requests
 // overrides same method from defaultIdentityServer struct
+// TODO need to introduce new liveness probe since node service doesn't discover AvailableCapacity anymore
```
Comment (mishoyama) [inline pkg/controller/controller.go:93]: Is this really required?

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 1 above)
Comment (mishoyama) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:130]: +1. Or use something like this:
```
switch {
	case isClean && (health == apiV1.HealthGood && status == apiV1.DriveStatusOnline):
        case !isClean:
		return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
	case health != apiV1.HealthGood || status != apiV1.DriveStatusOnline:
		return d.handleDriveIsNotGood(ctx, drive.Spec)	
	}
```

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+
+// createACIfNotExistOrUpdate tries to create AC for drive or update its size if AC already exists
+func (d *Controller) createACIfNotExistOrUpdate(ctx context.Context, drive api.Drive) (ctrl.Result, error) {
+	log := d.log.WithFields(logrus.Fields{
+		"method": "createACIfNotExistOrUpdate",
+	})
+	driveUUID := drive.GetUUID()
+	size := drive.GetSize()
+	// if drive is not clean, size is 0
+	if !drive.GetIsClean() {
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:146]: Won't work for LVG

File: pkg/base/linuxutils/fs/fs_helper.go (hunk 6 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/fs/fs_helper.go:331]: 2, 4 are detected in disocoverDataOnDrive function. 3 is implemented in this method. Isn't it enough to discover partitions? Why do we need to discover partition table?

File: pkg/controller/controller.go (hunk 12 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/controller.go:93]: Currently liveness and readiness probes depend on the readiness of Node pods. They become ready if Discover function is successful. When Node discovers all drives, ac and volumes, Controller starts working. With proposed implementation, Controller starts working even if not all AC are created. If it's ok for CSI, it is't required.

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 11 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:206]: It's the current implementation for unhealthy drives. I'll fix it then

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 8 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:109]: Yes, it's used for system LVG

File: cmd/controller/main.go (hunk 5 above)
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go:169]: We create cache `kubeCache` here and use it further in `cachedCrHelper` in `Controller` for drive and lvg

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 13 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:146]: Could you please add more details and information about the problem?

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 9 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:173]: `ErrorNotFound` is related to drive AC, not to LVG AC

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 10 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Fixed

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 7 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Fixed

File: cmd/controller/main.go (hunk 4 above)
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go]: Fixed

File: cmd/controller/main.go (hunk 3 above)
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go]: Fixed

File: pkg/node/volumemgr.go (hunk 2 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/node/volumemgr.go]: Fixed

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 1 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:129]: Fixed

File: pkg/base/linuxutils/fs/fs_helper.go (hunk 6 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/fs/fs_helper.go:331]: Implemented

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 14]
```diff
@@ -0,0 +1,378 @@
+package drivelvgcontroller
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:1]: name `drivelvgcontroller` sounds confusing. Can we have `capacitycontroller` instead?

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+	ac, err := d.cachedCrHelper.GetACByLocation(driveUUID)
+	switch {
+	case err == nil:
+		lvg, err := d.crHelper.GetLVGByDrive(ctx, driveUUID)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// If LVG exists for Drive, we delete drive AC, because CSI uses LVG AC
+		if lvg != nil {
+			if err := d.client.DeleteCR(context.WithValue(ctx, base.RequestUUID, ac.Name), ac); err != nil {
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:156]: We don't delete AC anymore. Set size to 0 instead.

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+			NodeId:       drive.GetNodeId(),
+		}
+		newAC := d.client.ConstructACCR(name, *capacity)
+		if err := d.client.CreateCR(context.WithValue(ctx, base.RequestUUID, name), name, newAC); err != nil {
+			log.Errorf("Error during create AvailableCapacity request to k8s: %v, error: %v",
+				capacity, err)
+			return ctrl.Result{}, err
+		}
+	default:
+		log.Infof("Failed to read AvailableCapacity for drive %s: %v", driveUUID, err)
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:189]: Why is this `INFO`?

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+	log := d.log.WithFields(logrus.Fields{
+		"method": "handleInaccessibleDrive",
+	})
+	ac, err := d.cachedCrHelper.GetACByLocation(drive.GetUUID())
+	switch {
+	case err == nil:
+		log.Infof("Update AC size to 0 %s based on unhealthy location %s", ac.Name, ac.Spec.Location)
+		ac.Spec.Size = 0
+		if err := d.client.UpdateCR(ctx, ac); err != nil {
+			log.Errorf("Failed to delete unhealthy available capacity CR: %v", err)
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:206]: Please update log message.

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+	switch {
+	case health != apiV1.HealthGood || status != apiV1.DriveStatusOnline:
+		return d.handleInaccessibleDrive(ctx, drive.Spec)
+	default:
+		return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
+	}
+}
+
+// createACIfNotExistOrUpdate tries to create AC for drive or update its size if AC already exists
+func (d *Controller) createACIfNotExistOrUpdate(ctx context.Context, drive api.Drive) (ctrl.Result, error) {
```
Comment (mishoyama) [inline pkg/controller/drivelvgcontroller/drivelvgcontroller.go:137]: Will name `createOrUpdateCapacity` sound better?

**Review by mishoyama — CHANGES_REQUESTED** [review]

File: cmd/controller/main.go [hunk 19]
```diff
@@ -142,23 +147,34 @@ func main() {
...
+
+	if err := accrd.AddToSchemeAvailableCapacity(scheme); err != nil {
+		return nil, err
+	}
+
+	if err := lvgcrd.AddToSchemeLVG(scheme); err != nil {
+		return nil, err
+	}
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go:174]: We can create 2 separate managers with different schemes for both controllers.

File: pkg/node/volumemgr.go [hunk 20]
```diff
@@ -743,174 +743,32 @@ func (m *VolumeManager) discoverVolumeCRs() error {
...
-		}
-		// confirm device SN if set
-		bDevSerialNum := bdev.Serial
-		driveSerialNum := drive.Spec.SerialNumber
-		if bDevSerialNum != "" && driveSerialNum != "" {
-			if !strings.EqualFold(bDevSerialNum, driveSerialNum) {
-				ll.Errorf("Serial numbers for block device %s don't match: %s != %s", bdev.Name, bDevSerialNum,
-					driveSerialNum)
-				continue
+		if _, ok := locations[drive.Spec.UUID]; ok {
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/node/volumemgr.go:752]: 1) I agree, we don't really need it as we discover it in `DiscoverData`
2) Yes

File: cmd/controller/main.go (hunk 5 above)
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go:189]: It's ok to use cache for AC, but not for LVG, so for LVG we use crHelper in capacity controller and for AC we use cacheCrHelper

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 15 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go:156]: This code doesn't relate to health change of Drive. This functionality was added to handle situation, when drive LVG is presented in cluster and drive AC exists. We should delete it as CSI uses LVG AC in this case. But I suppose this is not really probable situation, so this code can be removed.

File: pkg/base/linuxutils/fs/fs_helper.go [hunk 21]
```diff
@@ -53,7 +53,9 @@ const (
 	// RmDirCmdTmpl rm template
 	RmDirCmdTmpl = "rm -rf %s"
 	// WipeFSCmdTmpl cmd for wiping FS on device
-	WipeFSCmdTmpl = wipefs + "-af %s"
+	WipeFSCmdTmpl = wipefs + "-af %s" //
+	// DetectFSCmdTmpl cmd for detecting FS on device
+	DetectFSCmdTmpl = "lsblk %s --output FSTYPE --noheadings"
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/fs/fs_helper.go:58]: It seems like you check device partition. The idea of `DetectFs` is to use it for device without partition. If device has a partition, it's detected in `DeviceHasPartition` function.

File: cmd/controller/main.go [hunk 22]
```diff
@@ -109,19 +114,19 @@ func main() {
...
 		}()
 	}
-
+	stopCH := ctrl.SetupSignalHandler()
 	// todo make ACR feature mandatory and get rid of feature flag https://github.com/dell/csi-baremetal/issues/366
 	// check ACR feature flag
 	if featureConf.IsEnabled(featureconfig.FeatureACReservation) {
 		// create Reservation manager
-		reservationManager, err := createReservationManager(kubeClient, logger)
+		reservationManager, err := createManagers(kubeClient, logger, stopCH)
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go]: Fixed

File: cmd/controller/main.go [hunk 23]
```diff
@@ -109,19 +114,19 @@ func main() {
 			}
 		}()
 	}
-
+	stopCH := ctrl.SetupSignalHandler()
 	// todo make ACR feature mandatory and get rid of feature flag https://github.com/dell/csi-baremetal/issues/366
 	// check ACR feature flag
 	if featureConf.IsEnabled(featureconfig.FeatureACReservation) {
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go]: Done

File: cmd/controller/main.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply cmd/controller/main.go:167]: Scheme is used to serialize and deserialize object in kubernetes client, I suppose it controller doesn't use object, it doesn't affect whole performance

File: docs/proposals/ac-refactoring.md [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+## Abstract
+
+This proposal contains approaches for discovering of FS, LVG and partitions on drives and
+available capacity manipulations in CSI controller
+
+## Background
+
+Currently, we discover volumes on drive during `discoverVolumeCRs` function in volume manager and create `AvailableCapacity` 
+according to this information. Need to detect fs, lvm, partitions presence on disk, avoid creating new volume and move logic
+with AC manipulation to CSI controller to avoid possible race conditions.
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply docs/proposals/ac-refactoring.md]: Fixed

File: docs/proposals/ac-refactoring.md [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+## Proposal
+
+There are 3 possible approach to solve problem above.
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply docs/proposals/ac-refactoring.md]: Fixed

File: docs/proposals/ac-refactoring.md [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+1) Add Drive controller in CSI controller.
+2) Add `IsClean` field handling in reconcile loop. Controller update according AvailableCapacity with size 0.
+If drive is system, controller tries to update LVG AvailableCapacity.
+3) Add `Health` field handling in reconcile loop. If `Health` is not good, controller deletes AvailableCapacity and updates LVG AvailableCapacity.
+
+## Open issues (if applicable)
+
+ID | Name | Descriptions | Status | Comments
+---| -----| -------------| ------ | --------
+ISSUE-1 | How does Kubernetes handle 2 reconcile loop for one resource |  | Open |
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply docs/proposals/ac-refactoring.md]: Fixed

File: pkg/base/linuxutils/datadiscover/datadiscover.go [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// NewDataDiscover is a constructor for WrapDataDiscoverImpl
+func NewDataDiscover(fs fs.WrapFS,
+	part partitionhelper.WrapPartition,
+	lvm lvm.WrapLVM) *WrapDataDiscoverImpl {
+	return &WrapDataDiscoverImpl{fsHelper: fs, partHelper: part, lvmHelper: lvm}
+}
+
+// DiscoverData perform linux operation to determine if device has data on it
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/datadiscover/datadiscover.go]: Updated

File: pkg/base/linuxutils/fs/fs_helper.go [hunk 28]
```diff
@@ -312,3 +315,24 @@ func (h *WrapFSImpl) Unmount(path string) error {
...
+	var (
+		cmd          = fmt.Sprintf(DetectFSCmdTmpl, device)
+		stdout       string
+		driveHasData bool
+		err          error
+	)
+	if stdout, _, err = h.e.RunCmd(cmd,
+		command.UseMetrics(true),
+		command.CmdName(strings.TrimSpace(fmt.Sprintf(DetectFSCmdTmpl, "")))); err != nil {
+		return false, fmt.Errorf("failed to wipe file system on %s: %v", device, err)
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/fs/fs_helper.go]: Updated

File: pkg/base/linuxutils/fs/fs_helper.go [hunk 29]
```diff
@@ -312,3 +315,24 @@ func (h *WrapFSImpl) Unmount(path string) error {
...
+func (h *WrapFSImpl) DeviceHasFs(device string) (bool, error) {
+	var (
+		cmd          = fmt.Sprintf(DetectFSCmdTmpl, device)
+		stdout       string
+		driveHasData bool
+		err          error
+	)
+	if stdout, _, err = h.e.RunCmd(cmd,
+		command.UseMetrics(true),
+		command.CmdName(strings.TrimSpace(fmt.Sprintf(DetectFSCmdTmpl, "")))); err != nil {
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/fs/fs_helper.go]: Fixed

File: pkg/node/volumemgr.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/node/volumemgr.go:743]: Removed LVG check

File: pkg/base/linuxutils/lvm/lvm.go [hunk 30]
```diff
@@ -42,6 +42,8 @@ const (
 	PVRemoveCmdTmpl = lvmPath + "pvremove --yes %s" // add PV name
 	// PVsInVGCmdTmpl print PVs in VG cmd
 	PVsInVGCmdTmpl = lvmPath + "pvs --select vg_name=%s -o pv_name --noheadings" // add VG name
+	// VGsListCmdTmpl print all PVs name on node
+	VGsListCmdTmpl = lvmPath + "vgs -o pv_name --noheadings"
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/lvm/lvm.go]: Fixed

File: pkg/base/linuxutils/partitionhelper/partition_helper.go [hunk 31]
```diff
@@ -314,3 +321,61 @@ func (p *WrapPartitionImpl) GetPartitionNameByUUID(device, partUUID string) (str
...
+func (p *WrapPartitionImpl) DeviceHasPartitionTable(device string) (bool, error) {
+	/*
+		Disk /dev/sda: 931.5 GiB, 1000204886016 bytes, 1953525168 sectors
+		Units: sectors of 1 * 512 = 512 bytes
+		Sector size (logical/physical): 512 bytes / 512 bytes
+		I/O size (minimum/optimal): 512 bytes / 512 bytes
+		Disklabel type: gpt
+
+	*/
+	const labelKey = "Disklabel type"
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/base/linuxutils/partitionhelper/partition_helper.go]: Fixed

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 14 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Renamed

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 15 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Removed this code

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 16 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Fixed

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 17 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Updated

File: pkg/controller/drivelvgcontroller/drivelvgcontroller.go (hunk 18 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/controller/drivelvgcontroller/drivelvgcontroller.go]: Renamed

**Review by mishoyama — APPROVED** [review]

**Review by painhardcore — APPROVED** [review]

**Pull request merged**